pkg/controllers/volume_claim_policy.go

```diff
@@ -0,0 +1,122 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"maps"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+	"sigs.k8s.io/jobset/pkg/constants"
+)
+
+// reconcileVolumeClaimPolicies creates PVCs using volumeClaimPolicies.
+func (r *JobSetReconciler) reconcileVolumeClaimPolicies(ctx context.Context, js *jobset.JobSet) error {
+	var allErrors []error
+	// Create PVC for each volume claim policy.
+	for _, volumeClaimPolicy := range js.Spec.VolumeClaimPolicies {
+		if err := r.createPVCsIfNecessary(ctx, js, volumeClaimPolicy); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return errors.Join(allErrors...)
+}
+
+// createPVCsIfNecessary creates PVCs if it doesn't exist.
+func (r *JobSetReconciler) createPVCsIfNecessary(ctx context.Context, js *jobset.JobSet, volumeClaimPolicy jobset.VolumeClaimPolicy) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	for _, template := range volumeClaimPolicy.Templates {
+		// Add JobSet name label.
+		labels := make(map[string]string)
+		maps.Copy(labels, template.Labels)
+		labels[jobset.JobSetNameKey] = js.Name
+
+		// Create new PVC based on template.
+		pvc := corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        generatePVCName(js.Name, template.Name),
+				Namespace:   js.Namespace,
+				Labels:      labels,
+				Annotations: template.Annotations,
+			},
+			Spec: template.Spec,
+		}
+
+		// Set PVC owner reference if retention policy is Delete.
+		if volumeClaimPolicy.RetentionPolicy != nil && volumeClaimPolicy.RetentionPolicy.WhenDeleted == jobset.RetentionPolicyDelete {
+			if err := ctrl.SetControllerReference(js, &pvc, r.Scheme); err != nil {
+				return err
+			}
+		}
+
+		// Create PVC if it doesn't exist.
+		// TODO (andreyvelich): Shall we consider to use workqueue.ParallelizeUntil, similar to rJobs creation?
+		if err := r.Create(ctx, &pvc); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
```

Comment:
If retentionPolicy is Retain and someone creates a JobSet with a PVC.
And then they delete the JobSet with updated PVC templates will this cause any error?

Reply:
I guess this is a potential bug as PVC will be the old PVC.